src/read.jl

```diff
@@ -88,117 +84,108 @@ function read_header(io::IO)
-            n_bytes = 2 * signal.n_samples
-            data = Base.read(io, n_bytes)
-            if j == anno_idx
-                record = IOBuffer(data)
-                while !eof(record) && Base.peek(record) != 0x0
-                    toplevel, offset, duration, events = read_tal(record)
-                    if toplevel
-                        anno.offset = offset
-                        anno.event = events
-                        anno.n_bytes = n_bytes
-                    else
-                        push!(anno.annotations, AnnotationsList(offset, duration, events))
-                    end
-                end
-            else
-                append!(signal.samples, reinterpret(Int16, data))
-            end
-        end
-        push!(annos, anno)
+read_data!(samples::Vector{Int16}, data::Vector{UInt8}, ::SignalHeader) = append!(samples, reinterpret(Int16, data))
+
+function read_data!(samples::Vector{DataRecord}, data::Vector{UInt8}, ::AnnotationListHeader)
+    record = IOBuffer(data)
+    annotations = Vector{TimestampAnnotation}()
+    record_annotation = read_record_annotation(record)
+    while !eof(record) && Base.peek(record) != 0x0
+        push!(annotations, read_timestamp_annotation(record))
     end
-    deleteat!(signals, anno_idx)
-    @assert eof(io)
-    return (signals, annos)
+    if isempty(annotations)
+        annotations = nothing
+    end
+    push!(samples, record_annotation => annotations)
+    return nothing
 end
 
-function read_tal(io::IO)
-    c = Base.read(io, UInt8)
-    # Read the offset from the start time declared in the header
-    @assert c === 0x2b || c === 0x2d  # + or -
-    sign = c === 0x2b ? 1 : -1
-    buffer = UInt8[]
-    while true
-        c = Base.read(io, UInt8)
-        if c === 0x14 || c === 0x15
-            break
-        end
-        push!(buffer, c)
-    end
-    offset = sign * parse(Float64, String(buffer))
-    # Read the duration, if present
-    if c === 0x15
-        duration = parse(Float64, String(readuntil(io, 0x14)))
+samples(::SignalHeader) = Vector{Int16}()
+samples(::AnnotationListHeader) = Vector{DataRecord}()
+
+function read_record_annotation(io::IO)
+    sign = read_sign(io)
+    offset = sign * parse(Float64, String(readuntil(io, 0x14)))
+    if Base.peek(io) !== 0x0
+        events = read_events(io)
     else
-        duration = nothing
+        events = nothing
+        Base.skip(io, 1)
     end
-    c = Base.read(io, UInt8)
-    record = c === 0x14  # Whether the annotation applies to the entire record
-    # Read the annotation text, if present
-    if c !== 0x0
-        raw = readuntil(io, 0x0)
-        pushfirst!(raw, c)
-        events = convert(Vector{String}, split(String(raw), '\x14', keepempty=false))
+    return RecordAnnotation(offset, events)
+end
+
+function read_timestamp_annotation(io::IO)
+    sign = read_sign(io)
+    raw = readuntil(io, 0x14)
+    timestamp = split(String(raw), '\x15', keepempty=false)
```

Comment:
```suggestion
    timestamp = split(String(raw), '\x15'; keepempty=false)
```
See item 6 in https://github.com/jrevels/YASGuide#other-syntax-guidelines